jerry-core/ecma/builtin-objects/ecma-builtin-string-prototype.cpp

```diff
@@ -344,8 +347,245 @@ static ecma_completion_value_t
+                  ecma_op_check_object_coercible (this_arg),
+                  ret_value);
+
+  /* 2. */
+  ECMA_TRY_CATCH (this_to_string_value,
+                  ecma_op_to_string (this_arg),
+                  ret_value);
+
+  ecma_value_t regexp_value = 0;
+  /* 3. */
+  if (ecma_is_value_object (arg)
+      && ecma_object_get_class_name (ecma_get_object_from_value (arg)) == LIT_MAGIC_STRING_REGEXP_UL)
+  {
+    regexp_value = ecma_copy_value (arg, true);
+  }
+  else
+  {
+    /* 4. */
+    ecma_value_t regexp_arguments[1] = { arg };
+    ECMA_TRY_CATCH (new_regexp_value,
+                    ecma_builtin_regexp_dispatch_construct (regexp_arguments, 1),
+                    ret_value);
+
+    regexp_value = ecma_copy_value (new_regexp_value, true);
+
+    ECMA_FINALIZE (new_regexp_value);
+  }
+
+  if (ecma_is_completion_value_empty (ret_value))
+  {
+    JERRY_ASSERT (regexp_value != 0);
+    ecma_object_t *regexp_obj_p = ecma_get_object_from_value (regexp_value);
+    ecma_string_t *global_string_p = ecma_get_magic_string (LIT_MAGIC_STRING_GLOBAL);
+
+    /* 5. */
+    ECMA_TRY_CATCH (global_prop_value,
+                    ecma_op_object_get (regexp_obj_p, global_string_p),
+                    ret_value);
+
+    ECMA_TRY_CATCH (global_value,
+                    ecma_op_to_boolean (global_prop_value),
+                    ret_value);
+
+    ecma_value_t exec_arguments[1] = { this_to_string_value };
+
+    if (!ecma_is_completion_value_normal_true (global_value))
+    {
+      /* 7. */
+      ret_value = ecma_builtin_regexp_prototype_dispatch_routine (LIT_MAGIC_STRING_EXEC,
+                                                                  regexp_value,
+                                                                  exec_arguments,
+                                                                  1);
+    }
+    else
+    {
+      /* 8.a. */
+      ecma_number_t *zero_number_p = ecma_alloc_number ();
+      *zero_number_p = 0;
+
+      ecma_string_t *index_zero_string_p = ecma_new_ecma_string_from_uint32 (0);
+
+      ecma_string_t *last_index_string_p = ecma_get_magic_string (LIT_MAGIC_STRING_LASTINDEX_UL);
+
+      ECMA_TRY_CATCH (put_value,
+                      ecma_op_object_put (regexp_obj_p,
+                                          last_index_string_p,
+                                          ecma_make_number_value (zero_number_p),
+                                          true),
+                      ret_value);
+
+      /* 8.b. */
+      ECMA_TRY_CATCH (new_array_value,
+                      ecma_op_create_array_object (NULL, 0, false),
+                      ret_value);
+
+      ecma_object_t *new_array_obj_p = ecma_get_object_from_value (new_array_value);
+
+      /* 8.c. */
+      ecma_number_t previous_last_index = 0;
+      /* 8.d. */
+      uint32_t n = 0;
+      /* 8.e. */
+      bool last_match = true;
+
+      ecma_completion_value_t exec_result = ecma_make_empty_completion_value ();
```

Comment:
Why do we need the `exec_result` variable? Could we use `ret_value` instead of the variable?